File: event-sdk-generator/packages/runtime-typescript/src/adapters/sns.ts
import { SNSClient, PublishCommand } from "@aws-sdk/client-sns";
import type { TransportAdapter, EventEnvelope, PublishResult } from "../types.js";

export type SnsTransportAdapterOptions = {
  topicArn: string;
  region?: string;
  client?: SNSClient;
};

export class SnsTransportAdapter implements TransportAdapter {
  private readonly client: SNSClient;
  private readonly topicArn: string;

  constructor(options: SnsTransportAdapterOptions) {
    this.topicArn = options.topicArn;
    this.client = options.client ?? new SNSClient({ region: options.region ?? process.env["AWS_REGION"] ?? "us-east-1" });
  }

  async publish(envelope: EventEnvelope): Promise<PublishResult> {
    const fifo = envelope.metadata.fifo;
    const result = await this.client.send(
      new PublishCommand({
        TopicArn: this.topicArn,
        Message: JSON.stringify(envelope),
        MessageAttributes: {
          eventId: { DataType: "String", StringValue: envelope.eventId },
          version: { DataType: "String", StringValue: envelope.version },
        },
        ...(fifo && {
          MessageGroupId: fifo.messageGroupId,
          ...(fifo.deduplicationId && { MessageDeduplicationId: fifo.deduplicationId }),
        }),
      })
    );
    return { messageId: result.MessageId };
  }
}

File: packages/runtime-python/src/eventgen_runtime/adapters/sns.py
from __future__ import annotations
import json
import os
from eventgen_runtime.types import EventEnvelope, PublishResult


class SnsTransportAdapter:
    def __init__(self, topic_arn: str, region: str | None = None) -> None:
        self._topic_arn = topic_arn
        self._region = region or os.environ.get("AWS_REGION", "us-east-1")

    async def publish(self, envelope: EventEnvelope) -> PublishResult:
        import aioboto3  # type: ignore[import]
        session = aioboto3.Session()
        fifo = envelope.metadata.fifo
        extra: dict = {}
        if fifo:
            extra["MessageGroupId"] = fifo.message_group_id
            if fifo.deduplication_id:
                extra["MessageDeduplicationId"] = fifo.deduplication_id
        async with session.client("sns", region_name=self._region) as client:
            response = await client.publish(
                TopicArn=self._topic_arn,
                Message=json.dumps(envelope.__dict__, default=str),
                MessageAttributes={
                    "eventId": {"DataType": "String", "StringValue": envelope.event_id},
                    "version": {"DataType": "String", "StringValue": envelope.version},
                },
                **extra,
            )
        return PublishResult(message_id=response["MessageId"])
